src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -245,20 +245,45 @@ CitusBeginScanWithCoordinatorProcessing(CustomScanState *node, EState *estate, i
 
 	/* citus only evaluates functions for modification queries */
 	bool modifyQueryRequiresMasterEvaluation =
-		workerJob->requiresMasterEvaluation && jobQuery->commandType != CMD_SELECT;
+		jobQuery->commandType != CMD_SELECT &&
+		(workerJob->requiresMasterEvaluation || workerJob->deferredPruning);
 
 	/*
 	 * ExecuteMasterEvaluableFunctions handles both function evalation
 	 * and parameter evaluation. Pruning is most likely deferred because
 	 * there is a parameter on the distribution key. So, evaluate in both
 	 * cases.
 	 */
-	bool shoudEvaluteFunctionsOrParams =
-		modifyQueryRequiresMasterEvaluation || workerJob->deferredPruning;
-	if (shoudEvaluteFunctionsOrParams)
+	if (modifyQueryRequiresMasterEvaluation)
 	{
-		/* evaluate functions and parameters */
-		ExecuteMasterEvaluableFunctions(jobQuery, planState);
+		/* evaluate functions and parameters for modification queries */
+		ExecuteMasterEvaluableFunctionsAndParameters(jobQuery, planState);
+	}
+	else if (jobQuery->commandType == CMD_SELECT && !workerJob->deferredPruning)
+	{
+		/* we'll use generated strings, no need to have the parameters anymore */
+		EState *executorState = planState->state;
+		ResetExecutionParameters(executorState);
+
+		/* we're done,  we don't want to evaluate functions for SELECT queries */
+		return;
+	}
+	else if (jobQuery->commandType == CMD_SELECT && workerJob->deferredPruning)
+	{
+		/*
```

Comment:
This is my understanding on the reason, my confidence is medium-high, so please give feedback if the comment doesn't make sense  to you